### DISCUSSION TITLE
How to test Google OAuth flow using Cypress

### DISCUSSION THREAD

**Discussion by naripok** [supabase/supabase#9377] [Questions]
Hey guys!

I have a rather complex sign-in/sign-up/org-invitation use-case for my auth flow and I want to be able to test it e2e while covering the whole flow starting through any of the entry-points.

While there are ways to test a regular Google OAuth flow by using [Google's playground](https://developers.google.com/oauthplayground/) and some refresh tokens, I can see no way to test it in an automated fashion while using Supabase.

The problem is that supabase doesn't seem to use the regular google's refresh/access tokens, so the stuff I get from the playground is useless. If I try and use a testing framework like Cypress to click through the OAuth flow, I won't get consistent results as google will ask for additional validations (phone, email, captcha) from time to time. Google also has many different oauth consent pages that differ from region to region, browser to browser, if the browser is headless or not, etc, so it breaks the expectation for the prop selectors used in the browser testing.

So, the question here is is how to do it. How to test Supabase Google Oauth flow in an automated fashion without ending up with flaky tests.

Thanks!

**Comment by kraeki** [comment]
@naripok did you find a solution?

**Reply by naripok** [reply]
@kraeki Unfortunately, no.

We currently use a mix of going through the real app's google sign-in flow using playwright, getting the tokens set by supabase in the local storage from playwright's page, setting it in cypress' page `localStorage` directly so our app will assume the signed in state, and only then running the tests.

It is cumbersome, unpleasant, but this is the most reliable way we found to make it work.
Also, it is currently disabled on CI, as it generate some noise on ocasional failures.

**Comment by rookledookle** [comment]
Bumping this - trying to do the same thing but can't find a good solution

**Comment by ydennisy** [comment]
@kiwicopple this has been hanging around for quite some time and it would be great if someone from your team could take a look!

**Comment by hf** [comment]
> So, the question here is is how to do it. How to test Supabase Google Oauth flow in an automated fashion without ending up with flaky tests.

This is a limitation with Google's OAuth flow. It is intentionally designed to prevent bots -- like your Cypress tests -- from using it. Unfortunately, there's not much that we as Supabase can do for this.

**Reply by ydennisy** [reply]
@hf sorry but this is not strictly true, it has been done and cypress have a guide here: https://docs.cypress.io/guides/end-to-end-testing/google-authentication

I think it would be good to make a supabase specific guide for this.
